File: frogger_1d_recursive/frogger_1d.rb
def frog_hop(board, idx, n, m, h)
    move = board[idx]
    return "magic\n#{h}" if move == m
    return "cycle\n#{h}" if move == 0
    board[idx] = 0
    idx += move
    h += 1
    return "left\n#{h}" if idx < 0;
    return "right\n#{h}" if idx >= n
    frog_hop(board, idx, n, m, h)
end

n, s, m = gets.split.map(&:to_i)
h = 0
s -= 1
board = gets.split.map(&:to_i)
puts frog_hop(board, s, n, m, h)

File: frogger_1d_recursive/frogger_1d.py
def frog_hop(board, idx, n, m, h):
    move = board[idx]
    print(f"move = {move}")
    if move == m:
        return f"magic\n{h}"
    elif not move:
        return f"cycle\n{h}"
    board[idx] = 0
    idx += move
    h += 1
    if idx < 0:
        return f"left\n{h}"
    elif idx > n - 1:
        return f"right\n{h}"  
    return frog_hop(board, idx, n, m, h)

n, s, m = map(int, input().split())
h = 0
s -= 1
board = list(map(int, input().split()))
print(frog_hop(board, s, n, m, h))